### PR TITLE
Move Hive connector to transactional API

### PR DESCRIPTION
**Pull request by haozhun**

### CODE REVIEW

**Comment by electrum** [comment]
Move the `ClassLoaderSafe*` additions to a separate commit

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveConnectorFactory.java [hunk 1]
```diff
@@ -125,7 +123,8 @@ public Connector create(String connectorId, Map<String, String> config)
                     .initialize();
 
             LifeCycleManager lifeCycleManager = injector.getInstance(LifeCycleManager.class);
-            ConnectorMetadata metadata = injector.getInstance(ConnectorMetadata.class);
+            HiveMetadataFactory metadataFactory = injector.getInstance(HiveMetadataFactory.class);
+            //ConnectorMetadata metadata = injector.getInstance(ConnectorMetadata.class);
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveConnectorFactory.java]: Remove

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 2]
```diff
@@ -191,9 +160,9 @@ public HiveMetadata(
             LocationService locationService,
             TableParameterCodec tableParameterCodec,
             JsonCodec<PartitionUpdate> partitionUpdateCodec,
-            ExecutorService executorService)
+            BoundedExecutor renameExecution)
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: Make this type `ExecutorService`

Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: Name this `renameExecutor`

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 3]
```diff
@@ -609,10 +573,12 @@ public HiveOutputTableHandle beginCreateTable(ConnectorSession session, Connecto
                 partitionedBy,
                 tableMetadata.getOwner(),
                 additionalTableParameters);
+        setRollback(() -> rollbackCreateTable(result));
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: Add newline

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 4]
```diff
@@ -705,6 +674,8 @@ public HiveInsertTableHandle beginInsert(ConnectorSession session, ConnectorTabl
                 session.getQueryId(),
                 locationService.forExistingTable(session.getQueryId(), table.get()),
                 hiveStorageFormat);
+        setRollback(() -> rollbackInsert(result));
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: Newline

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadataFactory.java [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+import javax.inject.Inject;
+
+import java.util.concurrent.ExecutorService;
+import java.util.function.Supplier;
+
+import static java.util.Objects.requireNonNull;
+
+public class HiveMetadataFactory
+        implements Supplier<ConnectorMetadata>
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadataFactory.java]: Why is this a supplier?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadataFactory.java [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+                    timeZone.getID());
+        }
+
+        renameExecution = new BoundedExecutor(executorService, maxConcurrentFileRenames);
+    }
+
+    @Override
+    public ConnectorMetadata get()
+    {
+        return null;
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadataFactory.java]: This seems wrong

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientModule.java [hunk 7]
```diff
@@ -51,19 +50,22 @@
...
         this.typeManager = typeManager;
         this.pageIndexerFactory = pageIndexerFactory;
+        this.classLoader = classLoader;
     }
 
     @Override
     public void configure(Binder binder)
     {
         binder.bind(HiveConnectorId.class).toInstance(new HiveConnectorId(connectorId));
+        binder.bind(ClassLoader.class).toInstance(classLoader);
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveClientModule.java]: Where is this used? We shouldn't inject `ClassLoader`

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveConnector.java [hunk 8]
```diff
@@ -117,6 +133,39 @@ public ConnectorAccessControl getAccessControl()
...
+    public boolean isSingleStatementWritesOnly()
+    {
+        return true;
+    }
+
+    @Override
+    public ConnectorTransactionHandle beginTransaction(IsolationLevel isolationLevel, boolean readOnly)
+    {
+        checkConnectorSupports(READ_UNCOMMITTED, isolationLevel);
+        HiveTransactionHandle transaction = new HiveTransactionHandle();
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveConnector.java]: This variable can be `ConnectorTransactionHandle`

**Comment by dain** [comment]
One minor comment, but looks good

**Pull request merged**